### ISSUE TITLE
Personalized loss functions for tensor decompositions

### ISSUE THREAD

**Issue by earmingol** [tensorly/tensorly#486]
Any option to implement an easy way to "plug and play" different loss functions in the tensor decompositions?
It would be great having something like that instead of editing each of the different functions.

Tamara Kolda suggested more pertinent loss functions in a [talk she gave in 2018](https://www.youtube.com/watch?v=L8uT6hgMt00), depending on the type of data. For example, she suggested using the Rayleigh one for non-negative CP, or the Boolean-Odds for tensors with values 0s and 1s.

![Objective-Functions-Generalized-CP](https://user-images.githubusercontent.com/12451195/218014812-3e4857d6-3ce8-4f68-a9fd-efa947bf2b5c.png)

However, it looks like that most of the decomposition implemented here only use the standard loss function.

**Comment by JeanKossaifi** [comment]
If you want to use gradient-based optimization the easiest might be to just use [TensorLy-Torch](https://tensorly.org/torch/dev/) and plug in your loss directly.

**Comment by cohenjer** [comment]
So, it turns out we already have some implementation of GCP @JeanKossaifi @earmingol, but it is not yet merged in Tensorly mainly because we don't support all the backends. 

The implementation was done by @caglayantuna and the repo is here:
https://github.com/caglayantuna/tensorly-gcp
and can be pip installed [pypi link](https://pypi.org/project/tensorly-gcp/)
It has not been fully tested yet and this is something I want to properly finish but have not found the time for.

There are a few examples in the doc, hope this helps!

**Comment by earmingol** [comment]
Wow, this is cool! thanks @cohenjer, once I have a chance I'll start playing with it and see how I could contribute :)

@JeanKossaifi, I've got interested in trying tensorly-torch, but I'm not very familiar using it beyond creating layers for bigger models. Do you have any quick example that includes from passing the tensor input to creating the factor results?

Thanks!

**Comment by JeanKossaifi** [comment]
Yes, I need to add some documentation but it should be fairly straightforward -- I tried to make things as easy as possible :)

Create a [factorized tensor](https://tensorly.org/torch/dev/modules/generated/tltorch.FactorizedTensor.html#tltorch.FactorizedTensor) just like a PyTorch tensor 

```
from tltorch import FactorizedTensor

tucker = FactorizedTensor.new(shape=(3, 4, 2), rank=0.5, factorization='tucker')
```

You can use your favourite optimizer to update the factors of the decomposition
```
optimizer = torch.optim.Adam(
    tucker.parameters(), 
    lr=1e-3, 
    weight_decay=1e-4
)

```

Then you can just create a custom loss and apply it to the reconstruction:

```
reconstruction = tucker.to_tensor()
loss = custom_loss(reconstruction - target_tensor)
loss.backward()
```

You'd of course want to add in a scheduler to decrease learning rate every couple of iterations, etc.

**Comment by earmingol** [comment]
@JeanKossaifi unfortunately I got swamped by other things! I had a chance to try @cohenjer approach, which works well but outputs somehow unexpected results. Also, I tried implementing yours but wasn't able to get to the point to output results. That's something I need to do soon though!

I'll let you know if I have any progress or questions on that :)

**Comment by cohenjer** [comment]
@earmingol  Feel free to contact me directly if tensorly-gcp has issues. I plan to debug it at some point properly, so any feedback (even saying some functionality worked as expected) would be welcome!

**Comment by JeanKossaifi** [comment]
@earmingol are you still working on this or should we close the issue?
